src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,596 @@
+    std::lock_guard lock(mutex);
+    local_path_prefix = dir;
+    limit_size = limit_size_;
+
+    // scan local disk dir and recover the cache metas
+    Poco::File root_dir(local_path_prefix);
+    if (!root_dir.exists())
+        return;
+    auto callback = [this](RemoteCacheController * cntrl) { this->total_size += cntrl->size(); };
+
+    // four level dir. <schema>/<cluster>/<first 3 chars of path hash code>/<path hash code>``
+    std::vector<Poco::File> schema_dirs;
+    root_dir.list(schema_dirs);
+    for (auto & schema_file : schema_dirs)
+    {
+        std::vector<Poco::File> cluster_dirs;
+        schema_file.list(cluster_dirs);
+        for (auto & cluster_file : cluster_dirs)
+        {
+            std::vector<Poco::File> file_dir1;
+            cluster_file.list(file_dir1);
+            for (auto & file1 : file_dir1)
+            {
+                std::vector<Poco::File> file_dir2;
+                file1.list(file_dir2);
+                for (auto & file2 : file_dir2)
+                {
+                    if (caches.find(file2.path()) != caches.end())
+                    {
+                        LOG_ERROR(log, "ridiculous!! ");
+                        continue;
+                    }
+                    auto cache_cntrl = RemoteCacheController::recover(file2.path(), callback);
+                    if (cache_cntrl == nullptr)
+                        continue;
+                    CacheCell cell;
+                    cell.cache_controller = cache_cntrl;
+                    cell.key_iterator = keys.insert(keys.end(), file2.path());
+                    caches[file2.path()] = cell;
+                }
+            }
+        }
+    }
+    inited = true;
+}
+
+std::string
+RemoteReadBufferCache::calculateLocalPath(const std::string & schema_, const std::string & cluster_, const std::string & remote_path_)
+{
+    std::string local_path = local_path_prefix + "/" + schema_ + "/" + cluster_;
+
+    Poco::MD5Engine md5;
+    Poco::DigestOutputStream outstr(md5);
+    outstr << remote_path_;
+    outstr.flush(); //to pass everything to the digest engine
+    const Poco::DigestEngine::Digest & digest = md5.digest();
+    std::string md5string = Poco::DigestEngine::digestToHex(digest);
+
+    local_path += "/" + md5string.substr(0, 3) + "/" + md5string;
+
+    return local_path;
+}
+
+std::tuple<std::shared_ptr<LocalCachedFileReader>, RemoteReadBufferCache::CreateReaderError> RemoteReadBufferCache::createReader(
+    const std::string & schema,
+    const std::string & cluster,
+    const std::string & remote_path,
+    UInt64 mod_ts,
+    size_t file_size,
+    std::shared_ptr<ReadBuffer> & readbuffer)
+{
+    if (!hasInitialized())
+    {
+        LOG_ERROR(log, "RemoteReadBufferCache not init");
+        return {nullptr, CreateReaderError::NOT_INIT};
+    }
+    auto local_path = calculateLocalPath(schema, cluster, remote_path);
+    std::lock_guard lock(mutex);
+    auto citer = caches.find(local_path);
```

Comment:
Many non obvious names :(
What is citer? Please, make more understandable names of variables.

Reply:
done